# Step15 원-핫 인코딩과 교차 엔트로피 오차

2장은 목표값 t와 모델의 예측값 y 사이의 평균제곱오차(mse) 손실함수를 사용했다.(분류에서도 MSE를 사용가능 하긴 함)

하지만 여기서는 분류에서 주로 사용하는 손실함수를 설명한다.
1. 원-핫 인코딩
2. 카탈로그 교차 엔트로피 오차(CCE)
3. 이진 교차 엔트로피 오차(BCE)의

## 1. 원-핫 인코딩
원-핫 인코딩은 분류에서 목표값 레이블을 표현하는 방법 중 하나로,

각 카테고리에 대한 원-핫 인코딩은 부여받은 정수의 인덱스 위치만 1이고 나머지는 0을 갖는 것이다.

<정수 카테고리에 대한 원-핫 인코딩 이진 행렬을 반환>
```python
#1
tf.one_hot(indices, depth, ...) #indices의 정수를 depth 깊이로 원-핫 인코딩한 행렬의 텐서를 반환한다.
#2
tf.keras.utils.to_categorical(y, num_classes = None, dtype='float32')
#정수의 클래스 레이블 벡터(y)를 클래스의 개수(num_classes)로 원-핫 인코딩한 넘파이 행렬을 반환한다.
```
<br><br>
## step15_01
### 원-핫 인코딩: tf.one_hot(), tf.keras.utils.to_categorical()

In [3]:
import tensorflow as tf
import numpy as np

y = np.arange(10)
print("y=", y)

y1 = tf.keras.utils.to_categorical(y)
print("y1=", y1)

y2 = tf.one_hot(y, depth=10) #depth를 주어줘야한다.
print("y2=", y2.numpy())

y= [0 1 2 3 4 5 6 7 8 9]
y1= [[1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 1. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 1. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]]
y2= [[1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 1. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 1. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]]


## 2. 카테고리 교차 엔트로피 오차
p.114!!-> [수식 15.1], [수식 15.2]

카테고리 교차 엔트로피 함수는 여러 클래스 중에서 어느 하나의 클래스에 속지를 결정하는 다중 분류에 사용한다.

출력층은 클래스의 개수만큼의 뉴런을 가지며, 출력층의 활성화 함수는 softmax를 사용하고, 출력 뉴런 중에서 **최대 예측 값을 갖는 뉴런에 대한 인덱스의 클래스로 분류**한다.
<br><br>
1. 카테고리 교차 엔트로피: 원-핫 인코딩 목표값
```python
CCE = tf.keras.losses.CategoricalCrossentropy()
model.compile(optimizer='rmsprop', loss=CCE, metrics=['accuracy'])
model.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])
```
tf.keras.losses.CategoricalCrossentropy()는 다중 클래스 분류에서 **목표값이 원-핫 인코딩 되었을 때 교차 엔트로피 오차를 계산**한다.
<br><br><br>
2. 최소 카테고리 교차 엔트로피: 정수 레이블 인코딩
```python
CCE = tf.keras.losses.SparseCategoricalCrossentropy()
model.compile(optimizer='rmsprop', loss=CCE, metrics=['accuracy'])
model.compile(optimizer='rmsprop', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
```
tf.keras.losses.SparseCategoricalCrossentropy()는 다중 클래스 분류에서 **목표값이 정수 레이블로 인코딩 되었을 때 교차 엔트로피 오차를 계산**한다.

## step15_02
### tf.keras.losses.CategoricalCrossentropy()
1. 카테고리 교차 엔트로피: 원-핫 인코딩 목표값

In [12]:
import tensorflow as tf
import numpy as np

CCE = tf.keras.losses.CategoricalCrossentropy()
# t: 목표값, y: 원하는 출력값
t= np.array([[1,   0,   0,   0],   #t[0]
             [0,   1,   0,   0],   #t[1]
             [0,   0,   1,   0],   #t[2]
             [0,   0,   0,   1]])  #t[3]

y =np.array([[0.4, 0.3, 0.2, 0.1], #y[0]
             [0.1, 0.3, 0.2, 0.4]])#y[1]
#1
print("CCE(t[i], y[0])")
print("CCE(t[0], y[0])=", CCE(t[0], y[0]).numpy() ) 
print("CCE(t[1], y[0])=", CCE(t[1], y[0]).numpy() )
print("CCE(t[2], y[0])=", CCE(t[2], y[0]).numpy() )
print("CCE(t[3], y[0])=", CCE(t[3], y[0]).numpy() ) 

#2
print("CCE(t[i], y[1])")
print("CCE(t[0], y[1])=", CCE(t[0], y[1]).numpy() ) 
print("CCE(t[1], y[1])=", CCE(t[1], y[1]).numpy() )
print("CCE(t[2], y[1])=", CCE(t[2], y[1]).numpy() )
print("CCE(t[3], y[1])=", CCE(t[3], y[1]).numpy() )

#3
print("CCE(np.vstack((t[1], t[1])), y)=",
       CCE(np.vstack((t[1], t[1])), y).numpy() )

#이해 못하겠엉!
print(np.vstack((t[1],t[1])))

CCE(t[i], y[0])
CCE(t[0], y[0])= 0.916290731874155
CCE(t[1], y[0])= 1.203972804325936
CCE(t[2], y[0])= 1.6094379124341003
CCE(t[3], y[0])= 2.3025850929940455
CCE(t[i], y[1])
CCE(t[0], y[1])= 2.3025850929940455
CCE(t[1], y[1])= 1.203972804325936
CCE(t[2], y[1])= 1.6094379124341003
CCE(t[3], y[1])= 0.916290731874155
CCE(np.vstack((t[1], t[1])), y)= 1.203972804325936
[[0 1 0 0]
 [0 1 0 0]]


1. y[0]은 t[0](CCE(t[0], y[0])),
2. y[1]은 t[3](CCE(t[3], y[1]))과 교차엔트로피 오차가 제일 적다.

-> 2개의 미니배치 출력 y의 목표값이 np.vstack(t[1], t[1])일 때 교차 엔트로피 오차는 1.20이다.

즉, 교차 엔트로피의 평균 (CEE(y[0], t[0]) + CEE(y[1], t[1]))/2 이다.

<br><br>
-> y[0]은 t[0]과 오차가 가장 적고 y[1]은 t[3]과오차가 가작 작다.

## step15_02
### tf.keras.losses.SparseCategoricalCrossentropy()
2. 최소 카테고리 교차 엔트로피: 정수 레이블 인코딩

In [14]:
import tensorflow as tf
import numpy as np

SCE = tf.keras.losses.SparseCategoricalCrossentropy()

t = tf.convert_to_tensor([0, 1, 2, 3])
y =tf.convert_to_tensor([[0.4, 0.3, 0.2, 0.1], #y[0]
                         [0.1, 0.3, 0.2, 0.4]])#y[1]

#1
print("SCE(t[i], y[0])")
print("SCE(t[0], y[0])=", SCE(t[0], y[0]).numpy() ) 
print("SCE(t[1], y[0])=", SCE(t[1], y[0]).numpy() )
print("SCE(t[2], y[0])=", SCE(t[2], y[0]).numpy() )
print("SCE(t[3], y[0])=", SCE(t[3], y[0]).numpy() ) 

#2
print("SCE(t[i], y[1])")
print("SCE(t[0], y[1])=", SCE(t[0], y[1]).numpy() ) 
print("SCE(t[1], y[1])=", SCE(t[1], y[1]).numpy() )
print("SCE(t[2], y[1])=", SCE(t[2], y[1]).numpy() )
print("SCE(t[3], y[1])=", SCE(t[3], y[1]).numpy() )

#3
print("SCE(tf.stack((t[1], t[1])), y)=",
       SCE(tf.stack((t[1], t[1])), y).numpy() )

SCE(t[i], y[0])
SCE(t[0], y[0])= 0.91629076
SCE(t[1], y[0])= 1.2039728
SCE(t[2], y[0])= 1.609438
SCE(t[3], y[0])= 2.3025851
SCE(t[i], y[1])
SCE(t[0], y[1])= 2.3025851
SCE(t[1], y[1])= 1.2039728
SCE(t[2], y[1])= 1.609438
SCE(t[3], y[1])= 0.91629076
SCE(tf.stack((t[1], t[1])), y)= 1.2039728
1.4067054


2개의 미니배치 출력 y의 목표값이 np.vstack((t[1],t[1]))일 때, 교차 엔트로피 오차는 1.20이다.

즉, 교차 엔트로피의 평균'(SCE(t[1], y[0]) + SCE(t[1],y[1]))/2'이다.
<br><br>

-> y[0]은 t[0]과 오차가 가장 적고 y[1]은 t[3]과오차가 가작 작다.
<br><br><br>

## 3. 이진 교차 엔트로피 오차
p. 117: 수식 참고!

이진 교차 엔트로피 오차 손실함수는 출력층의 활성화 함수가 sigmoid를 사용할 때 주로 사용하며

훈련 데이터의 목표값에 출력층의 레이블을 속함(1), 속하지 않음(0)으로 표현한다.
<br><br>

1. 출력층이 1뉴런(유닛)이면 1-레이블을 분류할 수 있다. 각 훈련 데이터의 목표값은 0 또는 1로 표현한다.
2. 출력층이 2뉴런(유닛)이면 2-레이블을 분류할 수 있다. 각 훈련 데이터의 목표값은 [0,0], [0,1], [1,0], [1,1]로 4개의 클래스로 분류할 수 있다.

여러 개의 뉴런(유닛)을 갖는 출력층으로 다중 레이블 분류를 할 수 있다. 출력층의 각 뉴런의 예측값이 1에 가까운지, 0에 가까운지에 따라 존재를 판단한다.
```python
BCE = tf.keras.losses.BinaryCrossentropy()
model.compile(optimizer='rmsprop', loss=BCE, metrics=['accuracy'])
model.compile(optimizer=opt, loss='binary_crossentropy',metrics=['accruacy'])
```
tf.keras.losses.BinaryCrossentropy()는 다중 레이블 분류에서 이중 교차 엔트로피 오차를 계산한다.

## step15_04
### 이진 교차 엔트로피1: tf.keras.losses.BinaryCrossentropy()

In [19]:
import tensorflow as tf
import numpy as np

BCE = tf.keras.losses.BinaryCrossentropy()
t= np.array([[1,   0,   0,   0],   #t[0]
             [0,   1,   0,   0],   #t[1]
             [0,   0,   1,   0],   #t[2]
             [0,   0,   0,   1]])  #t[3]

y =np.array([[0.4, 0.3, 0.2, 0.1], #y[0]
             [0.1, 0.3, 0.2, 0.4]])#y[1]
#1
print("BCE(t[i], y[0])")
print("BCE(t[0], y[0])=", BCE(t[0], y[0]).numpy() )
print("BCE(t[1], y[0])=", BCE(t[1], y[0]).numpy() )
print("BCE(t[2], y[0])=", BCE(t[2], y[0]).numpy() )
print("BCE(t[3], y[0])=", BCE(t[3], y[0]).numpy() ) 

#2
print("BCE(t[i], y[1])")
print("BCE(t[0], y[1])=", BCE(t[0], y[1]).numpy() ) 
print("BCE(t[1], y[1])=", BCE(t[1], y[1]).numpy() )
print("BCE(t[2], y[1])=", BCE(t[2], y[1]).numpy() )
print("BCE(t[3], y[1])=", BCE(t[3], y[1]).numpy() )

#3
print("BCE(np.vstack((t[0], t[0])), y)=",
       BCE(np.vstack((t[0], t[0])), y).numpy() )


BCE(t[i], y[0])
BCE(t[0], y[0])= 0.4003672784541813
BCE(t[1], y[0])= 0.5108254397382338
BCE(t[2], y[0])= 0.6455745187904711
BCE(t[3], y[0])= 0.8483069443724252
BCE(t[i], y[1])
BCE(t[0], y[1])= 0.8483069443724252
BCE(t[1], y[1])= 0.5108254397382338
BCE(t[2], y[1])= 0.6455745187904711
BCE(t[3], y[1])= 0.40036727845418124
BCE(np.vstack((t[0], t[0])), y)= 0.6243371114133032


-> BCE(np.vstack((t[0], t[0])), y).numpy()가 0.62이다.

즉, 이진 교차 엔트로피의 평균은 (BCE(t[0],y[0]) + BCE(t[0],y[1]))/2이다.
<br><br>
-> y[0]은 t[0]과 오차가 가장 적고 y[1]은 t[3]과오차가 가작 작다.

## step15_05
### 이진 교차 엔트로피2: tf.keras.losses.BinaryCrossentropy()
step15_04 차이점 -> t가 이진 인코딩(원-핫 인코딩)이 아니다.

In [18]:
import tensorflow as tf
import numpy as np

BCE = tf.keras.losses.BinaryCrossentropy()
t= np.array([[1,   1,   0,   0],   #t[0]
             [0,   1,   1,   0],   #t[1]
             [0,   0,   1,   1],   #t[2]
             [0,   1,   0,   1]])  #t[3]

y =np.array([[0.4, 0.3, 0.2, 0.1], #y[0]
             [0.1, 0.3, 0.2, 0.4]])#y[1]
#1
print("BCE(t[i], y[0])")
print("BCE(t[0], y[0])=", BCE(t[0], y[0]).numpy() ) 
print("BCE(t[1], y[0])=", BCE(t[1], y[0]).numpy() )
print("BCE(t[2], y[0])=", BCE(t[2], y[0]).numpy() )
print("BCE(t[3], y[0])=", BCE(t[3], y[0]).numpy() ) 

#2
print("BCE(t[i], y[1])")
print("BCE(t[0], y[1])=", BCE(t[0], y[1]).numpy() ) 
print("BCE(t[1], y[1])=", BCE(t[1], y[1]).numpy() )
print("BCE(t[2], y[1])=", BCE(t[2], y[1]).numpy() )
print("BCE(t[3], y[1])=", BCE(t[3], y[1]).numpy() )

#3
print("BCE(np.vstack((t[0], t[0])), y)=",
       BCE(np.vstack((t[0], t[0])), y).numpy() )


BCE(t[i], y[0])
BCE(t[0], y[0])= 0.6121916959319459
BCE(t[1], y[0])= 0.8573989362682357
BCE(t[2], y[0])= 1.194880440902427
BCE(t[3], y[0])= 1.0601313618501897
BCE(t[i], y[1])
BCE(t[0], y[1])= 1.0601313618501897
BCE(t[1], y[1])= 0.8573989362682357
BCE(t[2], y[1])= 0.7469407749841832
BCE(t[3], y[1])= 0.6121916959319459
BCE(np.vstack((t[0], t[0])), y)= 0.8361615288910678


# Step16 활성화 함수
**p.122!!!**

일반적으로 뉴런의 출력은 y = f(WX + b)와 같이 활성화 함수(f())에 의해 뉴런의 출력을 제어한다.

<종류> 
1. linear: 입력 x를 그대로 출력한다.
2. sigmoid: (0, 1) 범위의 값으로 변환한다.
3. tanh: (-1, 1) 범위의 값으로 변환한다.
4. relu: 양수는 그대로, 음수는 0으로 변환한다.
5. LeakyReLU: 음수여도 alpha 값에 x를 곱해서 변환한다.
6. softmax: 지수함수를 사용하여 입력 벡터 x를 확률로 변환한다.

<사용법>

완전 연결층을 생성하는 Dense 층을 생성할 때, activation 인수에 함수 이름 또는 문자열('linear', 'sigmoid', 'tanh', 'softmax' 등)로 활성화 함수를 지정한다.

## step16_01
### 활성화 함수

In [21]:
import tensorflow as tf
import numpy as np

x = tf.constant([-10, -1.0, 0.0, 1.0, 10], dtype = tf.float32)

y1 = tf.keras.activations.linear(x) 
y2 = tf.keras.activations.sigmoid(x)
y3 = tf.keras.activations.tanh(x)
y4 = tf.keras.activations.relu(x)
y5 = tf.keras.layers.LeakyReLU(alpha=0.1)(x)
y6 = tf.keras.activations.softmax(tf.reshape(x, shape=(1, -1)))

##linear = tf.keras.activations.get('linear')
##y1 = linear(x)
##
##sigmoid = tf.keras.activations.get('sigmoid')
##y2 = sigmoid(x)
##
##tanh = tf.keras.activations.get('tanh')
##y3 = tanh(x)
##
##relu = tf.keras.activations.get('relu')
##y4 = relu(x)
##
##y5 = relu(x, alpha=0.1) # LeakyReLU
##softmax = tf.keras.activations.get('softmax')
##y6 = softmax(tf.reshape(x, shape=(1, -1)))

print("y1=", y1.numpy())
print("y2=", y2.numpy())
print("y3=", y3.numpy())
print("y4=", y4.numpy())
print("y5=", y5.numpy())
print("y6=", y6.numpy())
print("sum(y6)=", np.sum(y6.numpy())) # 1.0

y1= [-10.  -1.   0.   1.  10.]
y2= [4.5397868e-05 2.6894143e-01 5.0000000e-01 7.3105860e-01 9.9995458e-01]
y3= [-1.        -0.7615942  0.         0.7615942  1.       ]
y4= [ 0.  0.  0.  1. 10.]
y5= [-1.  -0.1  0.   1.  10. ]
y6= [[2.0607716e-09 1.6698603e-05 4.5391513e-05 1.2338691e-04 9.9981457e-01]]
sum(y6)= 1.0


# Step17 분류 성능평가
1. 정확도(accuracy): 전체에서 y_true, y_pred의 매칭 개수의 비율이다.
2. 정밀도(precision): Positive로 예측한 것 중에서 실제 Positive인 비율이다.
3. 재현율(recall): 실제 Positive인 것 중에서 Positive로 예측한 비율이다.
**P.123**
